File: client-sdks/python/anomaly_client.py
"""
AI Anomaly Detector Python Client SDK

Usage:
    from anomaly_client import AnomalyClient

    client = AnomalyClient('http://anomaly-detector:4000', api_key='your-key')

    # Send application metrics
    client.send_metric({
        'service': 'user-service',
        'endpoint': 'GET:/api/users',
        'metrics': {
            'response_time': 150,
            'status_code': 200
        }
    })

    # Send business metrics
    client.send_business_metric('daily_revenue', 45000, expected_range=[40000, 60000])

    # Send logs
    client.send_log('ERROR', 'Database timeout', 'payment-service')
"""

import json
import time
import requests
from typing import Dict, List, Optional, Union, Any
from datetime import datetime
import logging


class AnomalyClientError(Exception):
    """Custom exception for anomaly client errors"""

    pass


class AnomalyClient:
    """Python client for AI Anomaly Detector"""

    def __init__(self, base_url: str, api_key: Optional[str] = None, timeout: int = 30):
        """
        Initialize the anomaly client

        Args:
            base_url: Base URL of the anomaly detector service
            api_key: Optional API key for authentication
            timeout: Request timeout in seconds
        """
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout
        self.session = requests.Session()

        if api_key:
            self.session.headers.update({"Authorization": f"Bearer {api_key}"})

        self.session.headers.update(
            {
                "Content-Type": "application/json",
                "User-Agent": "AnomalyClient-Python/1.0",
            }
        )

        # Setup logging
        self.logger = logging.getLogger("AnomalyClient")

    def _make_request(
        self, method: str, endpoint: str, data: Optional[Dict] = None
    ) -> Dict:
        """Make HTTP request to the anomaly detector service"""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"

        try:
            response = self.session.request(
                method=method, url=url, json=data, timeout=self.timeout
            )

            if response.status_code >= 400:
                try:
                    error_data = response.json()
                    error_msg = error_data.get("error", f"HTTP {response.status_code}")
                except:
                    error_msg = f"HTTP {response.status_code}: {response.text}"

                raise AnomalyClientError(f"Request failed: {error_msg}")

            return response.json()

        except requests.RequestException as e:
            raise AnomalyClientError(f"Network error: {str(e)}")

    def health_check(self) -> Dict:
        """Check if the anomaly detector service is healthy"""
        return self._make_request("GET", "/health")

    def send_metric(self, metric_data: Dict) -> Dict:
        """
        Send application metrics

        Args:
            metric_data: Dictionary containing metric information
                Required fields: service, endpoint, metrics
                Optional fields: source_service, timestamp

        Example:
            client.send_metric({
                'service': 'user-service',
                'endpoint': 'GET:/api/users',
                'metrics': {
                    'response_time': 150,
                    'status_code': 200,
                    'payload_size': 1024
                },
                'source_service': 'web-gateway'
            })
        """
        if "timestamp" not in metric_data:
            metric_data["timestamp"] = datetime.utcnow().isoformat() + "Z"

        return self._make_request("POST", "/api/metrics", metric_data)

    def send_business_metric(
        self,
        metric_name: str,
        value: Union[int, float],
        expected_range: Optional[List[float]] = None,
        metadata: Optional[Dict] = None,
    ) -> Dict:
        """
        Send business metrics

        Args:
            metric_name: Name of the business metric
            value: Metric value
            expected_range: Optional expected range [min, max]
            metadata: Optional metadata dictionary

        Example:
            client.send_business_metric(
                'daily_revenue',
                45000,
                expected_range=[40000, 60000],
                metadata={'currency': 'USD', 'region': 'US-WEST'}
            )
        """
        data = {
            "metric_name": metric_name,
            "value": value,
            "timestamp": datetime.utcnow().isoformat() + "Z",
        }

        if expected_range:
            data["expected_range"] = expected_range

        if metadata:
            data["metadata"] = metadata

        return self._make_request("POST", "/api/business-metrics", data)

    def send_log(
        self, log_level: str, message: str, service: str, context: Optional[Dict] = None
    ) -> Dict:
        """
        Send log-based metrics

        Args:
            log_level: Log level (INFO, WARN, ERROR, FATAL, etc.)
            message: Log message
            service: Service name that generated the log
            context: Optional context dictionary

        Example:
            client.send_log(
                'ERROR',
                'Database connection timeout',
                'payment-service',
                context={'user_id': '12345', 'transaction_id': 'tx_789'}
            )
        """
        data = {
            "log_level": log_level.upper(),
            "message": message,
            "service": service,
            "timestamp": datetime.utcnow().isoformat() + "Z",
        }

        if context:
            data["context"] = context

        return self._make_request("POST", "/api/logs", data)

    def send_batch(self, metrics: List[Dict]) -> Dict:
        """
        Send multiple metrics in a single request

        Args:
            metrics: List of metric dictionaries

        Example:
            client.send_batch([
                {
                    'type': 'application',
                    'service': 'user-service',
                    'endpoint': 'GET:/api/users',
                    'metrics': {'response_time': 150, 'status_code': 200}
                },
                {
                    'type': 'business',
                    'metric_name': 'login_count',
                    'value': 1500
                }
            ])
        """
        data = {"metrics": metrics}
        return self._make_request("POST", "/api/batch", data)

    def get_config(self) -> Dict:
        """Get current anomaly detector configuration"""
        return self._make_request("GET", "/config")


class AnomalyMiddleware:
    """Middleware helper for automatic metric collection"""

    def __init__(self, client: AnomalyClient, service_name: str):
        self.client = client
        self.service_name = service_name

    def track_request(
        self,
        method: str,
        path: str,
        status_code: int,
        response_time: float,
        extra_metrics: Optional[Dict] = None,
    ):
        """Track a request automatically"""
        metrics = {"response_time": response_time, "status_code": status_code}

        if extra_metrics:
            metrics.update(extra_metrics)

        try:
            self.client.send_metric(
                {
                    "service": self.service_name,
                    "endpoint": f"{method}:{path}",
                    "metrics": metrics,
                }
            )
        except AnomalyClientError as e:
            # Don't let anomaly detection break the application
            logging.warning(f"Failed to send metrics: {e}")


# Context manager for automatic timing
class RequestTimer:
    """Context manager for timing requests"""

    def __init__(self, client: AnomalyClient, service: str, endpoint: str):
        self.client = client
        self.service = service
        self.endpoint = endpoint
        self.start_time = None

    def __enter__(self):
        self.start_time = time.time()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.start_time:
            duration = (time.time() - self.start_time) * 1000  # Convert to milliseconds
            status_code = 500 if exc_type else 200

            try:
                self.client.send_metric(
                    {
                        "service": self.service,
                        "endpoint": self.endpoint,
                        "metrics": {
                            "response_time": duration,
                            "status_code": status_code,
                        },
                    }
                )
            except AnomalyClientError:
                pass  # Silently fail to not break the application


# Decorator for automatic function timing
def track_function(client: AnomalyClient, service: str):
    """Decorator to automatically track function execution"""

    def decorator(func):
        def wrapper(*args, **kwargs):
            with RequestTimer(client, service, f"function:{func.__name__}"):
                return func(*args, **kwargs)

        return wrapper

    return decorator


# Example usage
if __name__ == "__main__":
    # Initialize client
    client = AnomalyClient("http://localhost:4000", api_key="your-api-key")

    # Check health
    try:
        health = client.health_check()
        print(f"Service health: {health}")
    except AnomalyClientError as e:
        print(f"Health check failed: {e}")

    # Send sample metrics
    try:
        # Application metric
        response = client.send_metric(
            {
                "service": "example-service",
                "endpoint": "GET:/api/test",
                "metrics": {
                    "response_time": 150,
                    "status_code": 200,
                    "payload_size": 1024,
                },
            }
        )
        print(f"Metric sent: {response}")

        # Business metric
        response = client.send_business_metric(
            "test_metric", 100, expected_range=[80, 120]
        )
        print(f"Business metric sent: {response}")

        # Log metric
        response = client.send_log("INFO", "Test log message", "example-service")
        print(f"Log sent: {response}")

    except AnomalyClientError as e:
        print(f"Error sending metrics: {e}")

File: client-sdks/javascript/anomaly-client.js
/**
 * AI Anomaly Detector JavaScript/Node.js Client SDK
 *
 * Usage:
 *   const AnomalyClient = require('./anomaly-client');
 *
 *   const client = new AnomalyClient('http://anomaly-detector:4000', 'your-api-key');
 *
 *   // Send application metrics
 *   await client.sendMetric({
 *     service: 'user-service',
 *     endpoint: 'GET:/api/users',
 *     metrics: {
 *       response_time: 150,
 *       status_code: 200
 *     }
 *   });
 *
 *   // Express.js middleware
 *   app.use(client.expressMiddleware('my-express-app'));
 */

const fetch = require("node-fetch");

class AnomalyClientError extends Error {
  constructor(message, statusCode = null) {
    super(message);
    this.name = "AnomalyClientError";
    this.statusCode = statusCode;
  }
}

class AnomalyClient {
  /**
   * Initialize the anomaly client
   * @param {string} baseUrl - Base URL of the anomaly detector service
   * @param {string} apiKey - Optional API key for authentication
   * @param {object} options - Additional options
   */
  constructor(baseUrl, apiKey = null, options = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.timeout = options.timeout || 30000;
    this.retries = options.retries || 3;

    this.headers = {
      "Content-Type": "application/json",
      "User-Agent": "AnomalyClient-JavaScript/1.0",
    };

    if (apiKey) {
      this.headers["Authorization"] = `Bearer ${apiKey}`;
    }
  }

  /**
   * Make HTTP request to the anomaly detector service
   * @param {string} method - HTTP method
   * @param {string} endpoint - API endpoint
   * @param {object} data - Request payload
   */
  async _makeRequest(method, endpoint, data = null) {
    const url = `${this.baseUrl}/${endpoint.replace(/^\//, "")}`;

    const options = {
      method,
      headers: this.headers,
      timeout: this.timeout,
    };

    if (data) {
      options.body = JSON.stringify(data);
    }

    let lastError;

    for (let attempt = 0; attempt < this.retries; attempt++) {
      try {
        const response = await fetch(url, options);

        if (!response.ok) {
          let errorMessage;
          try {
            const errorData = await response.json();
            errorMessage = errorData.error || `HTTP ${response.status}`;
          } catch {
            errorMessage = `HTTP ${response.status}: ${response.statusText}`;
          }
          throw new AnomalyClientError(errorMessage, response.status);
        }

        return await response.json();
      } catch (error) {
        lastError = error;
        if (attempt < this.retries - 1) {
          await this._sleep(Math.pow(2, attempt) * 1000); // Exponential backoff
        }
      }
    }

    throw new AnomalyClientError(
      `Request failed after ${this.retries} attempts: ${lastError.message}`
    );
  }

  /**
   * Sleep for specified milliseconds
   * @param {number} ms - Milliseconds to sleep
   */
  _sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  /**
   * Check if the anomaly detector service is healthy
   */
  async healthCheck() {
    return this._makeRequest("GET", "/health");
  }

  /**
   * Send application metrics
   * @param {object} metricData - Metric data object
   * @param {string} metricData.service - Service name
   * @param {string} metricData.endpoint - Endpoint identifier
   * @param {object} metricData.metrics - Metrics object
   * @param {string} [metricData.source_service] - Source service name
   * @param {string} [metricData.timestamp] - ISO timestamp
   */
  async sendMetric(metricData) {
    if (!metricData.timestamp) {
      metricData.timestamp = new Date().toISOString();
    }

    return this._makeRequest("POST", "/api/metrics", metricData);
  }

  /**
   * Send business metrics
   * @param {string} metricName - Name of the business metric
   * @param {number} value - Metric value
   * @param {Array<number>} [expectedRange] - Expected range [min, max]
   * @param {object} [metadata] - Additional metadata
   */
  async sendBusinessMetric(
    metricName,
    value,
    expectedRange = null,
    metadata = null
  ) {
    const data = {
      metric_name: metricName,
      value: value,
      timestamp: new Date().toISOString(),
    };

    if (expectedRange) {
      data.expected_range = expectedRange;
    }

    if (metadata) {
      data.metadata = metadata;
    }

    return this._makeRequest("POST", "/api/business-metrics", data);
  }

  /**
   * Send log-based metrics
   * @param {string} logLevel - Log level (INFO, WARN, ERROR, etc.)
   * @param {string} message - Log message
   * @param {string} service - Service name
   * @param {object} [context] - Additional context
   */
  async sendLog(logLevel, message, service, context = null) {
    const data = {
      log_level: logLevel.toUpperCase(),
      message: message,
      service: service,
      timestamp: new Date().toISOString(),
    };

    if (context) {
      data.context = context;
    }

    return this._makeRequest("POST", "/api/logs", data);
  }

  /**
   * Send multiple metrics in a single request
   * @param {Array<object>} metrics - Array of metric objects
   */
  async sendBatch(metrics) {
    return this._makeRequest("POST", "/api/batch", { metrics });
  }

  /**
   * Get current anomaly detector configuration
   */
  async getConfig() {
    return this._makeRequest("GET", "/config");
  }

  /**
   * Express.js middleware for automatic request tracking
   * @param {string} serviceName - Name of the service
   * @param {object} [options] - Middleware options
   */
  expressMiddleware(serviceName, options = {}) {
    const {
      trackErrors = true,
      trackSuccesses = true,
      extractUserId = null,
      ignoreRoutes = [],
    } = options;

    return (req, res, next) => {
      const startTime = Date.now();

      // Skip ignored routes
      if (ignoreRoutes.some((route) => req.path.includes(route))) {
        return next();
      }

      const originalSend = res.send;
      const originalJson = res.json;

      const sendMetrics = async () => {
        const duration = Date.now() - startTime;
        const shouldTrack =
          (res.statusCode >= 400 && trackErrors) ||
          (res.statusCode < 400 && trackSuccesses);

        if (!shouldTrack) return;

        const metrics = {
          response_time: duration,
          status_code: res.statusCode,
          payload_size: res.get("content-length") || 0,
        };

        // Extract user ID if function provided
        if (extractUserId && typeof extractUserId === "function") {
          try {
            const userId = extractUserId(req);
            if (userId) {
              metrics.user_id = userId;
            }
          } catch (error) {
            // Ignore extraction errors
          }
        }

        try {
          await this.sendMetric({
            service: serviceName,
            endpoint: `${req.method}:${req.route?.path || req.path}`,
            metrics: metrics,
            source_service: req.get("x-service-name") || "unknown",
          });
        } catch (error) {
          // Don't let anomaly detection break the application
          console.warn(
            "Failed to send metrics to anomaly detector:",
            error.message
          );
        }
      };

      // Override response methods to capture metrics
      res.send = function (body) {
        sendMetrics();
        return originalSend.call(this, body);
      };

      res.json = function (obj) {
        sendMetrics();
        return originalJson.call(this, obj);
      };

      next();
    };
  }

  /**
   * Koa.js middleware for automatic request tracking
   * @param {string} serviceName - Name of the service
   * @param {object} [options] - Middleware options
   */
  koaMiddleware(serviceName, options = {}) {
    const { trackErrors = true, trackSuccesses = true } = options;

    return async (ctx, next) => {
      const startTime = Date.now();

      try {
        await next();
      } finally {
        const duration = Date.now() - startTime;
        const shouldTrack =
          (ctx.status >= 400 && trackErrors) ||
          (ctx.status < 400 && trackSuccesses);

        if (shouldTrack) {
          try {
            await this.sendMetric({
              service: serviceName,
              endpoint: `${ctx.method}:${ctx.path}`,
              metrics: {
                response_time: duration,
                status_code: ctx.status,
                payload_size: ctx.length || 0,
              },
            });
          } catch (error) {
            console.warn(
              "Failed to send metrics to anomaly detector:",
              error.message
            );
          }
        }
      }
    };
  }
}

/**
 * Request timer helper class
 */
class RequestTimer {
  constructor(client, service, endpoint) {
    this.client = client;
    this.service = service;
    this.endpoint = endpoint;
    this.startTime = null;
  }

  start() {
    this.startTime = Date.now();
    return this;
  }

  async end(statusCode = 200, extraMetrics = {}) {
    if (!this.startTime) {
      throw new Error("Timer not started");
    }

    const duration = Date.now() - this.startTime;

    const metrics = {
      response_time: duration,
      status_code: statusCode,
      ...extraMetrics,
    };

    try {
      await this.client.sendMetric({
        service: this.service,
        endpoint: this.endpoint,
        metrics: metrics,
      });
    } catch (error) {
      console.warn("Failed to send metrics:", error.message);
    }
  }
}

/**
 * Function decorator for automatic timing
 * @param {AnomalyClient} client - Anomaly client instance
 * @param {string} service - Service name
 */
function trackFunction(client, service) {
  return function (target, propertyKey, descriptor) {
    const originalMethod = descriptor.value;

    descriptor.value = async function (...args) {
      const timer = new RequestTimer(
        client,
        service,
        `function:${propertyKey}`
      );
      timer.start();

      try {
        const result = await originalMethod.apply(this, args);
        await timer.end(200);
        return result;
      } catch (error) {
        await timer.end(500);
        throw error;
      }
    };

    return descriptor;
  };
}

// Export classes and utilities
module.exports = {
  AnomalyClient,
  AnomalyClientError,
  RequestTimer,
  trackFunction,
};

// Example usage
if (require.main === module) {
  async function example() {
    // Initialize client
    const client = new AnomalyClient("http://localhost:4000", "your-api-key");

    try {
      // Check health
      const health = await client.healthCheck();
      console.log("Service health:", health);

      // Send application metric
      const response = await client.sendMetric({
        service: "example-service",
        endpoint: "GET:/api/test",
        metrics: {
          response_time: 150,
          status_code: 200,
          payload_size: 1024,
        },
      });
      console.log("Metric sent:", response);

      // Send business metric
      const businessResponse = await client.sendBusinessMetric(
        "test_metric",
        100,
        [80, 120],
        { currency: "USD" }
      );
      console.log("Business metric sent:", businessResponse);

      // Send log
      const logResponse = await client.sendLog(
        "INFO",
        "Test log message",
        "example-service",
        { user_id: "12345" }
      );
      console.log("Log sent:", logResponse);
    } catch (error) {
      console.error("Error:", error.message);
    }
  }

  example();
}
